drivers/usb/device/usb_dc_rpi_pico.c

```diff
@@ -0,0 +1,819 @@
+{
+	struct udc_rpi_ep_state *ep_state = udc_rpi_get_ep_state(ep);
+
+	uint32_t val = len | USB_BUF_CTRL_AVAIL;
+
+	if (USB_EP_DIR_IS_IN(ep)) {
+		/* If writing to host, put data in place */
+
+		if (data) {
+			memcpy(ep_state->buf, data, len);
+			val |= USB_BUF_CTRL_FULL;
+		}
+	} else {
+		ep_state->read_offset = 0;
+
+		/* TODO: this seems to be needed to get the setup stage working,
+		 * otherwise I get DATA SEQ errors. This should be improved.
+		 */
+
+		if (USB_EP_GET_IDX(ep) == 0) {
+			ep_state->next_pid = 1;
+		}
+	}
+
+	LOG_DBG("xfer ep %d len %d pid: %d", ep, len, ep_state->next_pid);
+	val |= ep_state->next_pid ? USB_BUF_CTRL_DATA1_PID : USB_BUF_CTRL_DATA0_PID;
+
+	ep_state->next_pid ^= 1u;
+	*ep_state->buf_ctl = val;
+
+	return 0;
+}
+
+static void udc_rpi_handle_setup(void)
+{
+	struct cb_msg msg;
+
+	LOG_DBG("");
+
+	state.setup_available = true;
+
+	/* Reset PID to 1 for EP0 IN */
+	udc_rpi_get_ep_state(USB_CONTROL_EP_IN)->next_pid = 1;
+
+	msg.ep = USB_CONTROL_EP_OUT;
+	msg.cb = USB_DC_EP_SETUP;
+	msg.type = USB_DC_CB_TYPE_EP;
+
+	k_msgq_put(&usb_dc_msgq, &msg, K_NO_WAIT);
+}
+
+static void udc_rpi_handle_buff_status(void)
+{
+	struct udc_rpi_ep_state *ep;
+	enum usb_dc_ep_cb_status_code status_code;
+	uint8_t status = usb_hw->buf_status;
+	unsigned int i;
+	unsigned int bit = 1U;
+	struct cb_msg msg;
+
+	LOG_DBG("status: %d", status);
+
+	for (i = 0U; status && i < USB_NUM_BIDIR_ENDPOINTS * 2; i++) {
+		if (status & bit) {
+			usb_hw_clear->buf_status = bit;
+			bool in = !(i & 1U);
+			uint8_t ep_addr = (i >> 1U) | (in ? USB_EP_DIR_IN : USB_EP_DIR_OUT);
+
+			ep = udc_rpi_get_ep_state(ep_addr);
+			status_code = in ? USB_DC_EP_DATA_IN : USB_DC_EP_DATA_OUT;
+
+			LOG_DBG("buff ep %i in? %i", (i >> 1), in);
+
+			if (i == 0 && in && state.should_set_address) {
+				state.should_set_address = false;
+				usb_hw->dev_addr_ctrl = state.addr;
+			}
+
+			if (in) {
+				k_sem_give(&ep->write_sem);
+			}
+
+			msg.ep = ep_addr;
+			msg.type = USB_DC_CB_TYPE_EP;
+			msg.cb = status_code;
+
+			k_msgq_put(&usb_dc_msgq, &msg, K_NO_WAIT);
+
+			status &= ~bit;
+		}
+
+		bit <<= 1U;
+	}
+}
+
+static void udc_rpi_isr(const void *arg)
+{
+	uint32_t status = usb_hw->ints;
+	uint32_t handled = 0;
+	struct cb_msg msg;
+
+	if (status & USB_INTS_SETUP_REQ_BITS) {
+		handled |= USB_INTS_SETUP_REQ_BITS;
+		usb_hw_clear->sie_status = USB_SIE_STATUS_SETUP_REC_BITS;
+		udc_rpi_handle_setup();
+	}
+
+	if (status & USB_INTS_BUFF_STATUS_BITS) {
+		handled |= USB_INTS_BUFF_STATUS_BITS;
+		udc_rpi_handle_buff_status();
+	}
+
+	if (status & USB_INTS_DEV_CONN_DIS_BITS) {
+		LOG_DBG("buf %u ep %u", *udc_rpi_get_ep_state(0x81)->buf_ctl,
+			*udc_rpi_get_ep_state(0x81)->ep_ctl);
+		handled |= USB_INTS_DEV_CONN_DIS_BITS;
+		usb_hw_clear->sie_status = USB_SIE_STATUS_CONNECTED_BITS;
+
+		msg.ep = 0U;
+		msg.cb = usb_hw->sie_status & USB_SIE_STATUS_CONNECTED_BITS ?
+			USB_DC_CONNECTED :
+			USB_DC_DISCONNECTED;
```

Comment:
This, CONNECTED vs. DISCONNECTED, seems to be reversed.

Reply:
I'm not sure it is? Seems like the connected status bit being present should be `USB_DC_CONNECTED`.

Reply:
I observed it like this.

Reply:
Ok, I will swap, I haven't had a chance to test this myself directly.